presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryManager.java

```diff
@@ -240,14 +244,22 @@ public synchronized void process(Iterable<QueryExecution> runningQueries, Suppli
             return;
         }
 
-        // TODO revocable memory reservations can also leak and may need to be detected in the future
-        // We are only concerned about the leaks in general pool.
+        Map<QueryId, Optional<BasicQueryInfo>> localOrGloballyRunningQueries;
         if (memoryManagerService.isPresent()) {
-            memoryLeakDetector.checkForClusterMemoryLeaks(pools.get(GENERAL_POOL).getClusterInfo()::getRunningQueries, pools.get(GENERAL_POOL).getQueryMemoryReservations());
+            // We are in the multi-coordinator codepath, and thus care about the globally running queries
+            localOrGloballyRunningQueries = getClusterInfo(GENERAL_POOL)
+                            .getRunningQueries()
+                            .orElse(Collections.emptyList()).stream().collect(
+                                Collectors.toMap(Function.identity(), t -> Optional.empty()));
         }
         else {
-            memoryLeakDetector.checkForMemoryLeaks(allQueryInfoSupplier, pools.get(GENERAL_POOL).getQueryMemoryReservations());
+            // We are in the single coordinator setup, and thus care about the local queries. Ie, global queries
+            // does not make sense.
+            localOrGloballyRunningQueries = Maps.uniqueIndex(
+                    allQueryInfoSupplier.get().stream().map(queryInfo -> Optional.of(queryInfo)).iterator(),
+                    queryInfo -> queryInfo.get().getQueryId());
         }
+        memoryLeakDetector.checkForMemoryLeaks(localOrGloballyRunningQueries, pools.get(GENERAL_POOL).getQueryMemoryReservations());
```

Comment:
I feel the previous branching logic is much simpler than this one and easy to read. Can we keep the old way and just add the comments explaining the branching logic?

Reply:
I wanted to clarify this comment: Did you think that having two separate methods `checkForClusterMemoryLeaks` and  `checkForMemoryLeaks` in ClusterMemoryLeakDetector (like before) was simpler to follow ? 

My line of reasoning was to dedup this code because it was a bit subtle in terms of what all queries it considers and doesn't. 

But perhaps one middle ground could be to create a common helper method in the ClusterMemoryLeakDetector with two separate frontends. But the value add of that refactoring may not be much because we would be only calling those frontends (checkForClusterMemoryLeaks and checkForMemoryLeaks) in exactly one place each.

Reply:
it's fine, let's keep it as it is. 

> > > Also to avoid the flakyness of the metric, can we pull the memory leak detector logic outside and run it as a lesser frequncy (ie. once in 30 second/minute). This will help avoid us having fluctuation in metric. Today it runs every second which i feel is not necessary for memory leak detector logic. wdyt?
> > 
> > 
> > I agree but that sounds like a separate PR (this one is already creeping up). Unless you just want me to change the "1 second" to "30 seconds" (still hardcoded instead of configurable). But I didn't follow what you meant by "pull it outside" -- you mean that it should be outside the "process" method as in its thread and with a different period ?
> > I would be happy to make this change if you wouldn't have an issue with the PR size becoming big. I do agree that its important to not spam the log the every second.
> 
> Agree that it should be a seperate PR. Just feel it's unnecessary in the current flow where we keep looking for leaked queries at the same frequency where we enforcing some of the limits. It can be done in seperate thread at a much lower frequency than what it is today.

Also by having lesser frequency, we will have less flakyness in the leaked queries count that we can see with the change due to the frequent checks.

Reply:
PTAL at the current code. The change to pull out the leak check wasn't too bad -- I am happy to include it in this PR or can split it out.